### PR TITLE
interpret: use ScalarInt for bin-ops; avoid PartialOrd for ScalarInt

### PR DESCRIPTION
**Pull request by RalfJung**
Best reviewed commit-by-commit

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -6,9 +6,10 @@ use std::assert_matches::assert_matches;
 use either::{Either, Left, Right};
```
Comment (bjorn3) [inline compiler/rustc_const_eval/src/interpret/operand.rs:6]: Did you mean to say ints instead of its in the commit message of the first commit?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs:6]: Yes *oops*

**Comment by RalfJung** [comment]
Damn, I was a bit too optimistic.

**Comment by RalfJung** [comment]
That one is probably noise. So the tricky part is that pattern matching code. But I'm really not happy with `<=` on `ScalarInt`, that just doesn't have a clear semantics...

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 2]
```diff
@@ -242,6 +248,20 @@ impl<'tcx, Prov: Provenance> ImmTy<'tcx, Prov> {
         Self::from_scalar(Scalar::from_i8(c as i8), layout)
     }
 
+    /// Return the immediate as a `ScalarInt`. Ensures that it has the size that the layout of the
+    /// immediate indcates.
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/operand.rs:252]: indicates

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
Okay so that helped with perf but it is unfortunate that we can't use `try_to_scalar_int`... maybe there is a way.

@bors try @rust-timer queue

**Comment by RalfJung** [comment]
Nice. :)

This is ready. I wish github would tell me the commit ID at the time of your review, @oli-obk, then I could provide a diff...

**Review by oli-obk — APPROVED** [review]

**Pull request merged**